### ISSUE TITLE
Proper testing infrastructure

### ISSUE THREAD

**Issue by yannham** [tweag/nickel#161]
As the project codebase grows, more and more tests are accumulated. They are currently all mixed (unit, integration, regression, etc.), with sometimes hardcoded Nickel code, sometimes hand builded ASTs, and what not. This is not always easy to see what is actually covered and where. This is not the current priority until the MVP, but that could be nice to have a clean test suite and to detach generic language tests into separate nickel source files at some point.

**Comment by yannham** [comment]
Closing after #306 and #313. Tests for the stdlib are tracked in #321. There is of course a lot of room for improvement (property-based tests, generating lambda terms and testing against an reference implementation like Haskell/Dhall/Jsonnet, and so on), but the basic migrations have been carried out, and keeping this issue open would make its scope probably too broad and blurry.
